### ISSUE TITLE
Flakey tests in random modules

### ISSUE THREAD

**Issue by mihails-strasuns-sociomantic** [sociomantic-tsunami/ocean#204]
After we added code coverage checks, it started to complain sporadically about unexpected changes in few modules, most often `src/ocean/math/random/engines/KissCmwc.d`. Initially I dismissed it as either codecov or dmd bug, but recently I have realized that it is becomes those are tested with actual random seed and resulting coverage is not deterministic.

**Comment by mihails-strasuns-sociomantic** [comment]
Another source of flakey checks is `ISocket` - https://github.com/sociomantic-tsunami/ocean/blob/v2.x.x/src/ocean/sys/socket/model/ISocket.d#L139-L142 this condition may be true or not depending on some test timings.

**Comment by leandro-lucarella-sociomantic** [comment]
`src/ocean/math/random/engines/CMWC.d` seems to be another flakey module.

We really need to fix this, it's making coverage reports super annoying and unreliable.

**Comment by leandro-lucarella-sociomantic** [comment]
Probably anything calling `rSeed();` in the random package will be flakey.

**Comment by leandro-lucarella-sociomantic** [comment]
OK, I've been looking at this, and I found a dirty way to make the test coverage reproducible (for the `random` modules for now), but I'm not sure if it's the best approach. I just change a global seed in `ocean.math.random.Random` from a randomly generated number (taken from `/dev/random`) to a reproducible sequence of numbers.

The problem is the code seems to have test to check the seeds or some sort of randomicity (although, given that changing the seed to something reproduceable doesn't really make the tests fail, I wonder if I am misinterpreting the tests or if they are broken anyway).

Another option would be to try to run the random tests more than once to make it less likely to be flakey, although of course this option will never be fully reliable.

My take on this is, it's probably better to set the global seed to something reproducible every time that `-version=UnitTest` is used, as I guess any other modules using the seed could potentially have this issue too, and leave test to test the real randomicity to integration tests.

O the other hand, do we even use all these fancy random generators? I think using the generator based on `/dev/random` should be more than enough, so another option would be to just remove all other random number generators and that's it.

**Comment by leandro-lucarella-sociomantic** [comment]
This is one of the approaches... https://github.com/sociomantic-tsunami/ocean/pull/233

**Comment by mihails-strasuns-sociomantic** [comment]
> On the other hand, do we even use all these fancy random generators? 

I think most of code just uses whatever is default engine and it should be Kiss one.

**Comment by leandro-lucarella-sociomantic** [comment]
Kiss rather than udev that uses the hardware as a source of entropy? :thinking:

**Comment by mihails-strasuns-sociomantic** [comment]
https://github.com/sociomantic-tsunami/ocean/blob/v2.x.x/src/ocean/math/random/Random.d#L224-L226
and urandom is used to seed it - https://github.com/sociomantic-tsunami/ocean/blob/v2.x.x/src/ocean/math/random/Random.d#L1268-L1272

**Comment by leandro-lucarella-sociomantic** [comment]
Yeah, my question is why would anyone won't prefer to just use URandom for everything and that's it (in Linux)?

**Comment by mihails-strasuns-sociomantic** [comment]
Historical reasons :trollface:

**Comment by david-eckardt-sociomantic** [comment]
I'd be fine with creating one small module with a template function returning uniform `(u)int`, `(u)long` and `double` random numbers, based on `/dev/urandom` in reality and a predictable pseudo-random sequence for tests, and ban all other random number generators (including all I've ever used in the code I've written ;).

**Comment by leandro-lucarella-sociomantic** [comment]
Yeah, I was meaning as a way to move forward away from all the Tango bloat we inherited. We don't need that level of abstraction and options for random number generators. Although we should keep one reproducible pseudo-random generator for cases for unit tests and other corner cases where reproducibility might be desired.

**Comment by gavin-norman-sociomantic** [comment]
I would also be in favour of stripping things down to the basics that we actually need. (Another PR.)

**Comment by mihails-strasuns-sociomantic** [comment]
I don't mind though I have no idea how widely it is used right now and if some specific RNG requirements exist. @joseph-wakeling-sociomantic is probably the one who may be most picky about RNG stuff :)

**Comment by joseph-wakeling-sociomantic** [comment]
> My take on this is, it's probably better to set the global seed to something reproducible every time that -version=UnitTest is used, as I guess any other modules using the seed could potentially have this issue too, and leave test to test the real randomicity to integration tests.

Without looking into too much detail on this:

* if we have any sort of global RNG, it seems a VERY bad idea to me to re-seed it with anything deterministic, as this could interfere with other aspects of a downstream codebase's operations

* where deterministic tests are wanted for stuff using RNGs, it's probably best to have specific for-the-test-only RNG instances that are deterministically seeded

**Comment by joseph-wakeling-sociomantic** [comment]
BTW, I find the Tango RNG code to be a terrible mess anyway, so I'd be happy to see us move away from it to a better structured piece of work.

**Comment by joseph-wakeling-sociomantic** [comment]
Probably the best way to split things, BTW, is to separate out the functionality into:

* random _generators_, which generate (pseudo-)random sequences of uniformly distributed random bits

* random _distributions_, which transform random bits into random variates drawn from other statistical distributions

(this is C++11 standard terminology).

Then both components can be implemented on an as-needed basis.

(Contrast this with the horror of `ocean.math.random.Random` where the random generator is essentially an implementation detail underneath the hood, and the `Random` class is a monster megaclass trying to implement all the different random distributions you might want, a task in which it is inevitably going to fail.)

**Comment by mihails-strasuns-sociomantic** [comment]
@joseph-wakeling-sociomantic Question of what should be a proper RNG library is a bit more distant - right now I wanted your opinion about @leandro-lucarella-sociomantic proposal to kill all tango RNG engines completely and just hack `Random` to always use values straight from `/dev/urandom` - is there any chance it may be bad for optimization apps?

**Comment by joseph-wakeling-sociomantic** [comment]
I do not think you should use `/dev/urandom` under the hood.  First, it'll be a big speed hit (which might matter in some cases).  Second, it'll drain the system entropy pool unnecessarily.

`Random` is fairly implicitly meant to be pseudo-random by default, and I don't think we should mess with that.  It would be better to separate out the random distributions that we actually need/want and kill the `Random` class, rather than the pseudo-random engines.

**Comment by joseph-wakeling-sociomantic** [comment]
But the most important thing, as noted on @leandro-lucarella-sociomantic's patch:

I don't think that `version (UnitTest)` deterministic seeding is a viable solution for the problem, because it imposes a horrendous deterministic behaviour in a catch-all way. What if (for example) someone builds a program that uses random number generation and also uses the `-unittest` flag?

The correct solution is for unittests to seed the generator they use, for the scope of their tests, in a way that doesn't touch anyone else's random number generator.

**Comment by david-eckardt-sociomantic** [comment]
> I do not think you should use `/dev/urandom` under the hood. First, it'll be a big speed hit (which might matter in some cases). Second, it'll drain the system entropy pool unnecessarily.

I wonder. I understand your point, but especially when it comes to RNG entropy there are so many myths and misconceptions, and all statements I've read that actually reason about `urandom` rather than conveying hearsay conclude that one should just use it. I also prefer to question the claim of a performance impact until proven.

>  It would be better to separate out the random distributions that we actually need/want and kill the `Random` class, rather than the pseudo-random engines.

Yes. Any PRNG should be a struct with two public functions, one to seed it and one returning the next number.

> The correct solution is for unittests to seed the generator they use, for the scope of their tests, in a way that doesn't touch anyone else's random number generator.

I agree and actually [do that nowadays](https://github.com/sociomantic-tsunami/swarm/blob/0d06eb0355553c6bad6d8c259deca2e557114c68/src/swarm/neo/protocol/socket/MessageReceiver.d#L647).

**Comment by nemanja-boric-sociomantic** [comment]
>  all statements I've read that actually reason about urandom rather than conveying hearsay conclude that one should just use it

I believe (and I might be wrong) all these statements are made in context of `/dev/random` vs `/dev/urandom` (where the former can block, while the latter will use as much entropy for the RNG the system can provide and then just return values based on its current state). The general recommendation would be to prefer `/dev/urandom` over `/dev/random`, but not to prefer `/dev/urandom` over library PRNG).

It's certainly possible for the OS to need a real entropy for cryptographic purposes (and to use `/dev/random`) which will block as we've sucked all out via `/dev/urandom`.

> I also prefer to question the claim of a performance impact until proven.

Certainly it's clear that accessing `/dev/urandom` means _at least_ a context switch to a kernel space and it can affect performance not just because `/dev/urandom` might be slow, but because all the side-effects this can generate.

On top of this, only one thread in the system is allowed to read `/dev/urandom` concurrently, and on top of all other things, this is a deal-breaker IMHO: http://lkml.iu.edu//hypermail/linux/kernel/0412.1/0181.html


>This patch solves a problem where simultaneous reads to /dev/urandom can cause two processes on different processors to get the same value. We're not using a spinlock around the random generation loop because this will be a huge hit to preempt latency. So instead we just use a mutex around random_read and urandom_read. Yeah, it's not as efficient in the case of contention, if an application is calling /dev/urandom a huge amount, it's there's something really misdesigned with it, and we don't want to optimize for stupid applications.

**Comment by joseph-wakeling-sociomantic** [comment]
> I also prefer to question the claim of a performance impact until proven.

There's a very significant speed difference between `/dev/urandom` and almost all good pseudo-RNGs.  They are designed with very different purposes in mind and we shouldn't mix them up.

> I understand your point, but especially when it comes to RNG entropy there are so many myths and misconceptions, and all statements I've read that actually reason about urandom rather than conveying hearsay conclude that one should just use it.

`/dev/urandom` versus `/dev/random` discussions generally revolve around what is sufficiently non-predictable for the use-case in hand, and that's a completely separate discussion (where pseudo-RNGs are concerned, then one very rarely cares about whether the generator is predictable or not, one only cares about the statistical quality and the speed).

But in either case, one uses `/dev/urandom` or `/dev/random` where the fundamental unpredictability of the sequence of random bits is an important design consideration.  Where it isn't, one should generally avoid using those sources of randomness.

TL;DR: pseudo-RNGs have a fundamentally different use-case from something like `/dev/urandom` and we should preserve support for that different use-case.

**Comment by david-eckardt-sociomantic** [comment]
Fair enough, I'm not exactly an expert. But one good PRNG for (like the WELL or the Mersenne Twister) for everyone should be enough, shouldn't it? That would make three options:
1. `/dev/urandom` or [`getrandom()`](http://man7.org/linux/man-pages/man2/getrandom.2.html)
2. PRNG with local state
3. PRNG with global state, seeded on startup

This would still be a relatively small set compared to the forest of PRNGs in Tango, POSIX and C stdlib.

**Comment by joseph-wakeling-sociomantic** [comment]
> But one good PRNG for (like the WELL or the Mersenne Twister) for everyone should be enough, shouldn't it?

I'm reluctant to cut PRNGs that already exist.  There are a variety of use-cases and concerns out there (for example, Mersenne Twister has certain very nice properties which still make it useful in some situations, but OTOH there are generators which in other respects are statistically much better).

Are the different PRNGs really a maintenance burden?  Isn't the issue here how they have been used in unittests?

**Comment by joseph-wakeling-sociomantic** [comment]
> This would still be a relatively small set compared to the forest of PRNGs in Tango, POSIX and C stdlib.

I'm not sure what forest of PRNGs you have in mind in ocean (from our Tango legacy).  The total number is not large, and they are already written and surely don't need much in the way of maintenance?

The mere existence of multiple PRNGs can't in itself be responsible for the flaky-unittest issue described here.

**Comment by joseph-wakeling-sociomantic** [comment]
BTW, regarding:

> PRNG with global state, seeded on startup

... this should actually be a global instance of a PRNG.  Sorry to be picky, but it would be bad design to have a PRNG which inherently has global state.

(I suspect that you meant a global instance, but just to be clear.)

**Comment by david-eckardt-sociomantic** [comment]
> The mere existence of multiple PRNGs can't in itself be responsible for the flaky-unittest issue described here.

True, I was moving away from the issue, wondering why even several PRNGs are needed. But that's a different topic.

> I'm not sure what forest of PRNGs you have in mind in ocean (from our Tango legacy)

I just meant there are quite a couple in ocean/Tango and even more in C POSIX/stdlib. To a naive user like me this looks a bit excessive -- I'm sure one general-purpose PRNG would be enough. Again, this is not the issue here and a separate discussion.

**Comment by leandro-lucarella-sociomantic** [comment]
Generally I will just shut up because it's evident that I know nothing about RNG :)

But about this:
> I'm reluctant to cut PRNGs that already exist. There are a variety of use-cases and concerns out there (for example, Mersenne Twister has certain very nice properties which still make it useful in some situations, but OTOH there are generators which in other respects are statistically much better).

Ocean is supposed to be a general purpose library for very general tasks. If you have very specific needs in terms of RNG, I think it's fair to ask users to just use a RNG library instead of Tango.

**Comment by leandro-lucarella-sociomantic** [comment]
> The mere existence of multiple PRNGs can't in itself be responsible for the flaky-unittest issue described here.

The problem of keeping code that we don't use or understand is at sight right now. I just implemented a terrible solution to the flakey tests because I didn't want through the whole set of RNG and the random code that I don't understand and I don't want to spend time on.

If there was only one RNG that we really use, there would have been a better chance to get fixed properly insteady of trying to use a catch-all hack.

**Comment by joseph-wakeling-sociomantic** [comment]
@leandro-lucarella-sociomantic fair point.  I can see a case for reducing the number of supported pseudo-RNG engines, but I'm a bit concerned that it's sometimes useful to have at least a couple of pseudo-RNGs, because one can use one to generate seeds for another (I can go into detail if you want, but it's not so important).

However, on this comment from @mihails-strasuns-sociomantic:

> https://github.com/sociomantic-tsunami/ocean/blob/v2.x.x/src/ocean/math/random/Random.d#L224-L226
> and urandom is used to seed it - https://github.com/sociomantic-tsunami/ocean/blob/v2.x.x/src/ocean/math/random/Random.d#L1268-L1272

AFAICS that's a global instance of the default random number generator (or rather a `Random` based on the default engine).  It _should_ be unpredictably seeded because otherwise the onus is on the user to do this properly, and it _shouldn't_ be used in unittests (unless the unittest outcome is effectively independent from the particular random sequence generated).

Other than that, if you can point me at a few examples of flaky stuff in the PRNG modules, I'll take a deeper look at some point.  In general I don't think any of their unittests should include non-deterministic elements, so I'd be interested to see what unpredictable stuff is in there.

**Comment by mihails-strasuns-sociomantic** [comment]
Leaving one simple "I don't care PRNG" in ocean and moving others into separate dedicated library sounds like a good idea to me.

**Comment by mihails-strasuns-sociomantic** [comment]
I would propose to close this issue and create another one if you think we should plan to change RNG approach @leandro-lucarella-sociomantic

**Comment by leandro-lucarella-sociomantic** [comment]
It's still not completely clear to me what's the path we should take. If it's reasonable to use an external PRNG library, that would make sense to me, but I'm not sure if there is one that meet our needs out there. I would wait until we have a concrete plan, to create new issues.
